Docs on unitary synthesis (#420)
Qiskit/documentation: docs/build/unitary-synthesis.mdx

---

> title: Synthesizing unitary matrices

Reviewer: ```suggestion
title: Synthesize unitary matrices
```

> title: Synthesizing unitary matrices
> description: On the implementation of arbitrary unitary matrices on qubits
> ---
> 
> # Synthesizing unitaries
> 
> A unitary operation describes a norm-preserving change to a quantum system.
> For $n$ qubits this change is described by a $2^n \times 2^n$ dimensional, complex matrix $U$ whose adjoint equals the inverse, that is $U^\dagger U = \mathbb{1}$.

Reviewer: ```suggestion
title: Synthesizing unitary matrices
description: On the implementation of arbitrary unitary matrices on qubits
---

# Synthesize unitaries

A unitary operation describes a norm-preserving change to a quantum system.
For $n$ qubits this change is described by a $2^n \times 2^n$ dimensional, complex matrix $U$ whose adjoint equals the inverse; that is, $U^\dagger U = \mathbb{1}$.
```

Synthesizing specific unitary operations into a set of quantum gates is a fundamental task used, for example, in the design and application of quantum algorithms or in compiling quantum circuits.

While efficient synthesis is possible for certain classes of unitaries – like those composed of Clifford gates or having a tensor product structure – most unitaries do not fall into these categories.
For general unitary matrices, synthesis is a complex task with computational costs that increase exponentially with the number of qubits.
Therefore, if you know an efficient decomposition for the unitary you would like to implement, it will likely be better than a general synthesis.

<Admonition type="note">
    If no decomposition is available, Qiskit provides you with the tools to find one. 
    However, note that this generally generates deep circuits that may be unsuitable to run on noisy quantum computers.
</Admonition>

```python
import numpy as np
from qiskit import QuantumCircuit

U = 0.5 * np.array([
    [1, 1, 1, 1],
    [-1, 1, -1, 1],
    [-1, -1, 1, 1],
    [-1, 1, 1, -1]
])

circuit = QuantumCircuit(2)
circuit.unitary(U, circuit.qubits)
```

> ## Re-synthesis for circuit optimization
> 
> Sometimes it is beneficial to re-synthesis a long series of 1- and 2-qubit gates, if the length can be reduced. For example, the following circuit uses three 2-qubit gates.

Reviewer: ```suggestion
## Re-synthesis for circuit optimization

Sometimes it is beneficial to re-synthesize a long series of single- and two-qubit gates, if the length can be reduced. For example, the following circuit uses three two-qubit gates.
```

![output](/images/build/unitary-synthesis/unitary_target.png)

> However, after re-synthesizing it using the following code, it only needs a single CX gate.

Reviewer: ```suggestion
However, after re-synthesizing with the following code, it only needs a single CX gate.
```

```python 
from qiskit.quantum_info import Operator

# compute unitary matrix of target_circuit
U = Operator(target_circuit)

# re-synthesize
better_circuit = QuantumCircuit(2)
better_circuit.unitary(U, range(2))
better_circuit.decompose().draw()
```

![output](/images/build/unitary-synthesis/unitary_resynth.png)

> Qiskit's [transpile](../api/qiskit/compiler#qiskit.compiler.transpile) function automatically performs this re-synthesis for a sufficiently high optimization level,

Reviewer: ```suggestion
Qiskit's [transpile](../api/qiskit/compiler#qiskit.compiler.transpile) function automatically performs this re-synthesis for a sufficiently high optimization level.
```

> see also the [transpiler docs](../transpile/index).

Reviewer 2: ```suggestion
for more information about the Qiskit transpiler, take a look at the [transpile section](../transpile/index).
```

---
title: Synthesizing unitary matrices
description: On the implementation of arbirtary unitary matrices on qubits
---

> 

Reviewer 2: I think this is missing a title? It need a `# something`, otherwise it renders like this:
<img width="1382" alt="Screenshot 2023-11-28 at 15 27 10" src="https://github.com/Qiskit/documentation/assets/23662430/bfa5fe87-6514-4b03-98cb-457ce1b27d4b">

Author: Ah crap I thought the title in the `---` would be the header -- I'll add some

A unitary operation describes a norm-preserving change to a quantum system.

> For $n$ qubits is described by a $2^n \times 2^n$ dimensional, complex matrix $U$ whose adjoint equals the inverse, that is $U^\dagger U = \mathbb{1}$.

Reviewer 3: ```suggestion
For $n$ qubits this change is described by a $2^n \times 2^n$ dimensional, complex matrix $U$ whose adjoint equals the inverse, that is $U^\dagger U = \mathbb{1}$.
```

Synthesizing specific unitary operations into a set of quantum gates is a fundamental task used, for example, in the design and application of quantum algorithms or in compiling quantum circuits.

> While efficient synthesis is possible for certain classes of unitaries—like those composed of Clifford gates or having a tensor product structure—most unitaries do not fall into these categories.

Reviewer 3: ```suggestion
While efficient synthesis is possible for certain classes of unitaries – like those composed of Clifford gates or having a tensor product structure – most unitaries do not fall into these categories.
```

For general unitary matrices, synthesis is a complex task with computational costs that increase exponentially with the number of qubits.
Therefore, if you know an efficient decomposition for the unitary you would like to implement, it will likely be better than a general synthesis.

> If no decomposition is available Qiskit provides you with the tools to find one. 

Reviewer 3: ```suggestion
If no decomposition is available, Qiskit provides you with the tools to find one. 
```

> However, note that this generally generates deep circuits that may be unsuitable to run on noisy quantum computers.

Reviewer 2: I think this note could be put into a "Note" admonition box

Author: Do you have an example of that syntax? 🙂

```python
import numpy as np
from qiskit import QuantumCircuit

U = 0.5 * np.array([
    [1, 1, 1, 1],
    [-1, 1, -1, 1],
    [-1, -1, 1, 1],
    [-1, 1, 1, -1]
])

circuit = QuantumCircuit(2)
circuit.unitary(U, circuit.qubits)
```

> Sometimes it is beneficial to re-synthesis a long series of 1- and 2-qubit gates, if the length can be reduced. For example the circuit

Reviewer 3: ```suggestion
Sometimes it is beneficial to re-synthesis a long series of 1- and 2-qubit gates, if the length can be reduced. For example, the following circuit uses three 2-qubit gates.
```

I don't think we're supposed to split sentences with code blocks

> Sometimes it is beneficial to re-synthesis a long series of 1- and 2-qubit gates, if the length can be reduced. For example the circuit

Reviewer 2: Instead of spreading the code snippets out within one sentence could you write multiple sentences? It feels a bit jarring to cut the sentences off with multiple code blocks between them

```

> >>> target_circuit.draw()

Reviewer 2: could you draw this in the mpl way? I think thats the standard we're using for circuit diagrams for the other pages

Author: Is that possible in markdown or do we need to use a Jupyter notebook instead?

     ┌───┐              ┌───┐     ┌───┐
q_0: ┤ H ├──■─────────■─┤ X ├──■──┤ H ├
     └───┘┌─┴─┐┌────┐ │ ├───┤┌─┴─┐└───┘
q_1: ─────┤ X ├┤ √X ├─■─┤ X ├┤ X ├─────
          └───┘└────┘   └───┘└───┘
```

> is currently using three 2-qubit gates. However, after re-synthesizing it via

Reviewer 3: ```suggestion
However, after re-synthesizing it using the following code, it only needs a single CX gate.
```

```python 
from qiskit.quantum_info import Operator

# compute unitary matrix of target_circuit
U = Operator(target_circuit)

# re-synthesize
better_circuit = QuantumCircuit(2)
better_circuit.unitary(U, range(2))
```

> only needs a single CX gate

Reviewer 3: ```suggestion
```

```
>>> better_circuit.decompose().draw()

global phase: 3.2522
        ┌────────────┐           ┌────────────┐
q_0: ───┤ U(π/2,0,0) ├─────■─────┤ U(π/2,0,0) ├───
     ┌──┴────────────┴──┐┌─┴─┐┌──┴────────────┴──┐
q_1: ┤ U(1.3496,π/2,-π) ├┤ X ├┤ U(π/2,2.9204,-π) ├
     └──────────────────┘└───┘└──────────────────┘
```

> Qiskit's [transpile](../api/qiskit/compiler#qiskit.compiler.transpile) function automatically performs this re-synthesis for a sufficiently high optimization level.

Reviewer 2: maybe link to the new transpile section?
